include/envoy/config/typed_config.h

```diff
@@ -0,0 +1,60 @@
+#pragma once
+
+#include "envoy/common/pure.h"
+
+#include "common/protobuf/protobuf.h"
+
+namespace Envoy {
+namespace Config {
+
+/**
+ * Base class for an extension factory.
+ */
+class UntypedFactory {
+public:
+  virtual ~UntypedFactory() = default;
+
+  /**
+   * Name of the factory, a reversed DNS name is encouraged to avoid cross-org conflict.
+   * It's used as key in the metadata map, as well as key in the factory registry.
+   */
+  virtual const std::string name() const PURE;
+
+  /**
+   * @return std::string the identifying category name for objects
+   * created by this factory. Used for automatic registration with
+   * FactoryCategoryRegistry.
+   */
+  virtual const std::string category() const PURE;
+
+  /**
+   * @return configuration proto full name, or empty for untyped factories.
+   */
+  virtual const std::string configType() { return ""; }
+};
+
+/**
+ * Base class for an extension factory configured by a typed proto message.
+ */
+class TypedFactory : public UntypedFactory {
+public:
+  virtual ~TypedFactory() = default;
+
+  /**
+   * @return ProtobufTypes::MessagePtr create empty config proto message for v2. The config, which
+   * arrives in an opaque google.protobuf.Struct message, will be converted to JSON and then parsed
+   * into this empty proto.
+   */
+  virtual ProtobufTypes::MessagePtr createEmptyConfigProto() PURE;
```

Comment:
One thing we could do, and it doesn't have to be this PR, is move away from having each individual filter return empty protos. Instead, they could either return a type name or descriptor, and we could have more general machinery for assembling the empty proto. Might not be a big win, but just thought I'd throw it out there.

Reply:
Agree. It also helps with the registration initialization fiasco, since protobuf initializer may run after factory initializers.